### DISCUSSION TITLE
How to deal with errors after payment goes through

### DISCUSSION THREAD

**Discussion by vikitorfg** [pay-rails/pay#564] [Q&A] [answered]
I'm using jumpstart and paygem, after ```payment_processor.charge(@cart.total)``` runs and no errors are raised the customer I assume the card is charged charged. 
Is there a best way to deal with error **after** the charge. Let's say for some reason @order.save returns an error. Is there an error I can raise for ```payment_processor``` object that will undo the current charge? 

I could run ```.refund!``` before returning the error, but I feel like there could be a better option that I can't find. 
Or what the best business logic practices are when dealing with this kind of thing (firing an email to the administrator to get the order fixed for example). 

Below my code for some reference. Thanks a lot in advance.

```
def create
    payment_processor = params[:processor] ? current_account.set_payment_processor(params[:processor]) : current_account.payment_processor
    payment_processor.update_payment_method(params[:payment_method_token]) unless params[:processor].nil?
    payment_processor.charge(@cart.total)

    # This is after payment is confirmed, so the codes are created and delivered
    @order = Order.new(account_id: current_account.id, pay_charge_id: payment_processor.id)

    if @order.save
      conversion = @order.convert(@cart)
    else
      # TODO error handling
      # Raise error
      # Redirect_to error page
    end

    @order.order_items.each do |item|
      Rcode.create_bulk_rcodes(item.quantity, current_account, item.region_course_id, @order.id, payment_processor.id)
    end

  rescue Pay::ActionRequired => e
    redirect_to pay.payment_path(e.payment.id)
  rescue Pay::Error => e
    flash[:alert] = e.message
    render :new, status: :unprocessable_entity
  end
```

**Comment by excid3** [comment, marked as answer]
The first thing I would do is run `@order.valid?` before creating the charge. That way you can pre-verify the order is going save correctly before attempting the charge.

If there is a failure even after that, you may want to refund the charge or email an admin to handle it manually.
